presto-main/src/main/java/com/facebook/presto/operator/scalar/CharacterStringCasts.java

```diff
@@ -106,35 +106,28 @@ public static Slice varcharToCharSaturatedFloorCast(@LiteralParameter("y") long
          * We decrement last character in input (in fact, we decrement last non-zero character) and pad the value with
          * max code point up to y characters.
          */
-        int lastNonZeroCharIndex = codePoints.size() - 1;
-        while (lastNonZeroCharIndex >= 0 && codePoints.get(lastNonZeroCharIndex) == 0) {
-            codePoints.remove(lastNonZeroCharIndex);
-            lastNonZeroCharIndex--;
-        }
+        codePoints = trimTrailing(codePoints, '\0');
 
-        if (lastNonZeroCharIndex < 0) {
+        if (codePoints.isEmpty()) {
             // No non-zero characters in input and input is shorter than y. Input value is smaller than any char(4) casted back to varchar, so we return the smallest char(4) possible
-            return codePointsToSliceUtf8(nCopies(toIntExact(y), 0));
+            return codePointsToSliceUtf8(nCopies(toIntExact(y), (int) '\0'));
         }
 
-        codePoints.set(lastNonZeroCharIndex, codePoints.get(lastNonZeroCharIndex) - 1);
-        verify(codePoints.size() == lastNonZeroCharIndex + 1);
+        codePoints.set(codePoints.size() - 1, codePoints.get(codePoints.size() - 1) - 1);
         codePoints.addAll(nCopies(toIntExact(y) - codePoints.size(), Character.MAX_CODE_POINT));
 
-        verify(!codePoints.isEmpty());
         verify(codePoints.get(codePoints.size() - 1) != ' '); // no trailing spaces to trim
 
         return codePointsToSliceUtf8(codePoints);
     }
 
-    private static int trimTrailingSpaces(List<Integer> codePoints)
+    private static List<Integer> trimTrailing(List<Integer> codePoints, int codePointToTrim)
     {
-        int numberOfTrailingSpaces = 0;
-        while (!codePoints.isEmpty() && codePoints.get(codePoints.size() - 1) == ' ') {
-            codePoints.remove(codePoints.size() - 1);
-            numberOfTrailingSpaces--;
+        List<Integer> trimmed = new ArrayList<>(codePoints);
+        while (!trimmed.isEmpty() && trimmed.get(trimmed.size() - 1) == codePointToTrim) {
```

Comment:
You could construct empty list. Compute the length of prefix to be copied from source list. And the perform copy instead adding all elements and then removing suffix.
THough I am fine with this approach too.